### ISSUE TITLE
Support for fwd-declaring T when it only appears inside std::unique_ptr<T>

### ISSUE THREAD

**Issue by t-moe** [include-what-you-use/include-what-you-use#1297]
It seems that iwyu currently includes the definitions of T, even when T is only used as part of unique_ptr.

E.g. if I have a class containing a member `std::unique_ptr<Bar>` iwu will happily include bar.h, even if Bar is not used otherwise.

**Comment by bolshakov-a** [comment]
Does your class have an explicit destructor? IWYU ensures callability of implicitly generated default constructor and destructor. And `std::unique_ptr::~unique_ptr()` requires the full info for pointed-to type. When the destructor is defined in `.cpp` file, IWYU should require full `T` type in `.cpp` and not for the class definition in a header.

**Comment by t-moe** [comment]
Yes, our classes use explicit (non default) destructors, defined in the header file and implemented in the cpp.

Unless I'm mistaken include what you use suggests including the header file for the pointed-to type in that case.

-------- Original-Nachricht --------
Am 10. Sept. 2023, 20:00, schrieb Andrey Ali Khan Bolshakov :

**Comment by bolshakov-a** [comment]
Could you please send a reproducer?

**Comment by bolshakov-a** [comment]
And what implementation of the standard library do you use?

**Comment by t-moe** [comment]
I was using clang16  and the libstd coming with it (all from the ubuntu22.04 repos).

Unfortunately, I can't send you a reproducer, because I no longer have access to this environment. I'm sorry and I'll 
 try to attach a minimal, reproducible example right away.


What I take from this conversation is, that in theory include-what-you-use should support such a usecase. So we can close this issue for now and I will reopen another one, should I run into the issue again.

_____

If I remember correctly the code was something like this:

container.h
```

// include-what-you-use suggested to include item.h here
class Item;

class Container
{
    
    ~Container();
private:
     std::unique_ptr<Item> _item;
};
```

container.cpp
```
#include "container.h"
Container::~Container(){}
```


item.h

```
class Item {};
```

**Comment by bolshakov-a** [comment]
I cannot reproduce it both with libc++-16 and libstdc++-12, neither on the top of the master branch nor at the 0.20 tag. Assuming `<memory>` and `"item.h"` are `#include`d in `container.h`, IWYU output is:
```
container.h should add these lines:

container.h should remove these lines:
- #include "item.h"  // lines 2-2

The full include-list for container.h:
#include <memory>  // for unique_ptr
class Item;  // lines 4-4
---

container.cpp should add these lines:
#include "item.h"  // for Item

container.cpp should remove these lines:

The full include-list for container.cpp:
#include "container.h"
#include "item.h"  // for Item

```

**Comment by kimgr** [comment]
Given bolshakov-a's repro results above, I'd say this is fixed on mainline.

**Comment by bolshakov-a** [comment]
It turns out that the complete `Item` type is required not only for `Container` destructor but also for a constructor definition:
> In a non-delegating constructor, the destructor for each potentially constructed subobject of class type is [potentially invoked](https://timsong-cpp.github.io/cppwp/n4659/class.dtor#def:potentially_invoked). [ Note: This provision ensures that destructors can be called for fully-constructed subobjects in case an exception is thrown ([[except.ctor]](https://timsong-cpp.github.io/cppwp/n4659/except.ctor)).  — end note ]

https://timsong-cpp.github.io/cppwp/n4659/class.base.init#12

So a bug is that IWYU does NOT require the full `Item` type in `container.h`. It should do that unless a `Container` constructor is declared (not implicitly defined) there because it would assure that you can define a `Container` variable (if the ctor and dtor were `public`, of course; in the example above the dtor is private).
